File: Advanced Robotics Course/Chapter17/Lesson1/Chapter17_Lesson1.m
N = 10;
m = 0.05;
k_spring = 50;
c_damp = 0.1;
g = 9.81;

% Mass and stiffness for 1D vertical motion per node (y-coordinate only)
M = m * eye(N);
K = zeros(N);
for i = 1:N-1
    e = zeros(N,1);
    e(i) = 1; e(i+1) = -1;
    K = K + k_spring * (e * e');
end

% Damping proportional to mass
C = c_damp * eye(N);

dt = 5e-4;
steps = 20000;
y = zeros(N,1);
v = zeros(N,1);

% Initialization (rope horizontal, small sag)
for i = 1:N
    y(i) = 0.0;
end

for step = 1:steps
    % Gravity
    f_ext = -g * m * ones(N,1);

    % Anchor first node
    f_ext(1) = 0;
    y(1) = 0;
    v(1) = 0;

    % Drive last node (robot gripper)
    t = step * dt;
    y(N) = 0.1 * sin(2*pi*t);
    v(N) = 0;

    % Internal elastic and damping forces: M * a = f_ext - C*v - K*y
    rhs = f_ext - C * v - K * y;
    a = M \ rhs;

    % Semi-implicit Euler
    v = v + dt * a;
    y = y + dt * v;
end

% In Simulink, put M, C, K into a second-order state-space block and
% connect the last node to the robot gripper trajectory as an input.
      

File: Advanced Robotics Course/Chapter17/Lesson1/Chapter17_Lesson1.py
import numpy as np

# Number of nodes in the rope
N = 10
mass = 0.05
k_spring = 50.0
damping = 0.1
g = 9.81

# Small time step due to stiffness
dt = 5e-4
steps = 20000

# Positions and velocities in 2D
x = np.zeros((N, 2))
v = np.zeros_like(x)

# Initialize rope horizontally
rest_length = 0.05
for i in range(N):
    x[i, 0] = rest_length * i
    x[i, 1] = 0.0

rest_lengths = np.full(N - 1, rest_length)

def compute_forces(x, v):
    f = np.zeros_like(x)

    # Internal springs between neighbors
    for i in range(N - 1):
        dx = x[i + 1] - x[i]
        L = np.linalg.norm(dx)
        if L > 1e-6:
            direction = dx / L
            f_spring = k_spring * (L - rest_lengths[i]) * direction
            f[i] += f_spring
            f[i + 1] -= f_spring

    # Damping and gravity
    f -= damping * v
    f[:, 1] -= g * mass
    return f

for step in range(steps):
    # External forces
    f = compute_forces(x, v)

    # Fix the first node (anchored to the table)
    f[0, :] = 0.0
    v[0, :] = 0.0
    x[0, :] = np.array([0.0, 0.0])

    # Drive the last node along a simple trajectory (robot gripper)
    t = step * dt
    target = np.array([rest_length * (N - 1),
                       0.1 * np.sin(2.0 * np.pi * t)])
    x[-1, :] = target
    v[-1, :] = np.array([0.0, 0.0])

    # Semi-implicit (symplectic) Euler integration
    a = f / mass
    v += dt * a
    x += dt * v

# In practice, this core integration loop would be embedded in a robotics
# simulator (e.g., PyBullet, MuJoCo), and coupled to robot joint-space control.
      
